tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter_test.cc

```diff
@@ -631,6 +642,46 @@ TEST_F(CudnnConvRewriterTest, BackwardInputConvolveConstantFilter) {
                           0));
 }
 
+// Check that an integer forward convolution is replaced with a custom call to
+// cudnnConvolutionForward
+TEST_F(CudnnConvRewriterTest, ForwardIntegerConvolution) {
+  const string module_str = absl::StrFormat(R"(
+    HloModule Test
+
+    ENTRY Test {
+      input = s8[1,17,3,3] parameter(0)
+      filter = s8[3,3,17,63] parameter(1)
+
+      ROOT conv = f32[1,63,3,3] convolution(input, filter), window={size=3x3 pad=1_1x1_1}, dim_labels=bf01_01io->bf01, feature_group_count=1
+    })");
+  TF_ASSERT_OK_AND_ASSIGN(auto m, ParseAndReturnVerifiedModule(module_str));
+
+  EXPECT_TRUE(RunPass(m.get()));
+  EXPECT_THAT(
+      m->entry_computation()->root_instruction(),
+      op::GetTupleElement(
+          op::CustomCall(kCudnnConvForwardCallTarget, _, op::Parameter()), 0));
+
+  string optimized_hlo_string = GetOptimizedHlo(module_str);
+  EXPECT_TRUE(RunAndCompare(module_str, ErrorSpec{0.01}))
+      << optimized_hlo_string;
+}
+
+// Check that an integer forward convolution is not allowed
+TEST_F(CudnnConvRewriterTest, ForwardIntegerOutputConvolution) {
+  const string module_str = absl::StrFormat(R"(
+    HloModule Test
+
+    ENTRY Test {
+      input = s8[1,2,3,3] parameter(0)
+      filter = s8[3,3,2,5] parameter(1)
+
+      ROOT conv = s8[1,5,3,3] convolution(input, filter), window={size=3x3 pad=1_1x1_1}, dim_labels=bf01_01io->bf01, feature_group_count=1
+    })");
+  TF_ASSERT_OK_AND_ASSIGN(auto m, ParseAndReturnVerifiedModule(module_str));
+
+  //   ASSERT_FALSE(CudnnConvRewriter().Run(m.get()).ValueOrDie());
```

Comment:
I think it's sufficient to `ASSERT_FALSE(CudnnConvRewriter().Run(m.get()).ok())`.

Reply:
I tried the recommended change.  Although I could see the expected error message in the test log, the test didn't finish properly.

[ RUN      ] CudnnConvRewriterTest.ForwardIntegerOutputConvolution
2019-07-18 05:59:31.063159: I tensorflow/compiler/xla/service/platform_util.cc:78] platform Host present but no XLA compiler available: could not find registered compiler for platform Host -- check target linkage (hint: try linking in tensorflow/compiler/jit:xla_cpu_jit)
2019-07-18 05:59:31.063200: I tensorflow/compiler/xla/service/platform_util.cc:78] platform CUDA present but no XLA compiler available: could not find registered compiler for platform CUDA -- check target linkage (hint: try linking in tensorflow/compiler/jit:xla_gpu_jit)
2019-07-18 05:59:31.063232: I tensorflow/compiler/xla/service/platform_util.cc:78] platform Host present but no XLA compiler available: could not find registered compiler for platform Host -- check target linkage (hint: try linking in tensorflow/compiler/jit:xla_cpu_jit)
2019-07-18 05:59:31.063246: I tensorflow/compiler/xla/service/platform_util.cc:78] platform CUDA present but no XLA compiler available: could not find registered compiler for platform CUDA -- check target linkage (hint: try linking in tensorflow/compiler/jit:xla_gpu_jit)
2019-07-18 05:59:31.068303: F tensorflow/stream_executor/lib/statusor.cc:34] Attempting to fetch value instead of handling error Unimplemented: The convolution instruction with integer inputs only allows float outputs.  Insert a clamp instruction with range [-128, 127) followed by a convert instruction after the convolution instruction for int8 outputs.
external/bazel_tools/tools/test/test-setup.sh: line 310:   549 Aborted                 (core dumped) "${TEST_PATH}" "$@" 2>&1

Reply:
Have you tried linking `//tensorflow/compiler/jit:xla_gpu_jit` to the test target?

Reply:
What do you mean linking //tensorflow/compiler/jit:xla_gpu_jit to the test target?  How to do that?  The test doesn't specify dependency on xla_gpu_jit.

Is it because I use ().ValueOrDie() instead of passing status back to the call stack?

Reply:
XLA defines several Bazel targets as "plugins". They are cc_libraries that the final binary (e.g. a test target) may choose to depend on or not. For GPU plugin, if the test doesn't depend on it, all GPU related code is not linked into the final binary. The error message is saying sort of that.

I think the word "linking" is confusing, but it means to add //tensorflow/compiler/jit:xla_gpu_jit to the test target as a dependency.

Reply:
Yes, "linking //tensorflow/compiler/jit:xla_gpu_jit" makes the test working.  I have made another commit for the changes.

I saw the message about "linking" in the log and considered it information message.  Is GPU plugin required by error handling, but not by regular HLO pattern matching tests.